File: 프로그래머스/2/154538. 숫자 변환하기/숫자 변환하기.java
import java.util.*;
class Solution {
    public int solution(int x, int y, int n) {
        int[] dp = new int[y+1];
        for (int i = 0 ; i <= y; i++) dp[i] = -1;
        dp[x] = 0;
        Queue<int[]> q = new ArrayDeque<>();
        q.add(new int[]{x,0});
        while(!q.isEmpty()) {
            int[] cur = q.poll();
            int value = cur[0];
            int step = cur[1];
            if (value * 3 <= y && dp[value*3] == -1) {
                q.add(new int[] {value*3,step+1});
                dp[value*3] = step+1;
            }
            if (value * 2 <= y && dp[value*2] == -1) {
                q.add(new int[]{value*2,step+1});
                dp[value*2] = step+1;
            }
            if (value + n <= y && dp[value+n] == -1) {
                q.add(new int[]{value+n,step+1});
                dp[value+n] = step+1;
            }
        }
        return dp[y];
    }
}

File: 프로그래머스/2/154538. 숫자 변환하기/숫자 변환하기.kt
import java.util.ArrayDeque
class Solution {
    fun solution(x: Int, y: Int, n: Int): Int {
        val q = ArrayDeque<Pair<Int,Int>>()
        val visit = Array(y+1) { false }
        visit[x] = true
        q.addLast(Pair(x,0))
        while (q.isNotEmpty()) {
            val (cur,cnt) = q.removeFirst()
            if (cur == y) return cnt
            if (cur + n <= y && !visit[cur+n]) {
                q.addLast(Pair(cur+n,cnt+1))   
                visit[cur+n] = true
            }
            if (cur * 2 <= y) {
                q.addLast(Pair(cur*2,cnt+1))
                visit[cur*2] = true
            }
            if (cur * 3 <= y) {
                q.addLast(Pair(cur*3,cnt+1))
                visit[cur*3] = true
            }
        }
        return -1
    }
}